### PR TITLE
Bug fixing and more tests for LAPACK

### PR DESCRIPTION
**Pull request by kshyatt**
Some methods in `base/linalg/lapack.jl` were completely untested and had type bugs! I've fixed a few, and added more argument-related error throwing. I also added a bunch of tests for untested error throws and totally untested methods. I improved a test for `posv` that was occasionally tossing `PosDefException` on my local machine.

### CODE REVIEW

**Comment by tkelman** [comment]
Did you catch the typo from https://github.com/JuliaLang/julia/commit/363678a6d396ebe5afaf99369a3d706997d40ebe#commitcomment-11698552

(@ScottPJones it's generally better to comment in a PR than on the merge commit, commit comments are easy to lose track of over time)

**Comment by kshyatt** [comment]
I didn't! Thanks to @tkelman for the reminder, and @ScottPJones for the initial catch! I've updated the PR.

**Comment by kshyatt** [comment]
Aaaaand owned by whitespace. Not a good day for me, clearly.

**Comment by ScottPJones** [comment]
Thanks, @tkelman, I wasn't aware of that (I look forward to the day when I don't feel like such a newbie here!)
@kshyatt Not a good day for you?  You accomplish quite a lot of good, then, even on a bad day!
BTW, what do you think of my ideas for error handling (which I used in `utferror.jl`)?

File: test/linalg/lapack.jl [hunk 1]
```diff
@@ -54,6 +54,83 @@ let # Issue #7886
...
+    B = copy(A)
+    C,T = LAPACK.geqrt!(A,zeros(elty,10,10))
+    D,S = LAPACK.geqrt3!(A,zeros(elty,10,10))
+    @test_approx_eq C D
+end
+
+#geqp3, geqrt, error handling!
+for elty in (Float32, Float64, Complex64, Complex128)
+    A = rand(elty,10,10)
+    @test_throws DimensionMismatch LAPACK.geqp3!(A,ones(Int,11),zeros(elty,10))
```
Comment (andreasnoack) [inline test/linalg/lapack.jl]: These should be `BlasInt` and must be the reason for the 32 bit test failures.

**Comment by andreasnoack** [comment]
Looks great. Only problem I saw was the `Int`s.

File: test/linalg/lapack.jl (hunk 1 above)
Comment (kshyatt) [inline test/linalg/lapack.jl]: This explains why it passed locally... what an embarrassing PR!

**Comment by ScottPJones** [comment]
@kshyatt I mentioned before, that I'd been told to use the more julian construct:
```julia
errorcondition && throw(ErrorType(arguments))
```
instead of
```julia
if errorcondition
      throw(ErrorType(arguments)
end
```
But then your changes brought up the issue that the first idiom, used very heavily throughout the code,
does not work well with the way the coverage tools don't really handle conditional code on a single line.
I was wondering this morning, in the interest of more concise code, if your nice error checking could be written as:
```julia
errorcondition &&
    throw(ErrorType(arguments))
```
(BTW, for some reason the code quoting is messing up the indentation... I wrote that with the standard for julia 4 space indentation)

**Comment by tkelman** [comment]
Would need to test whether the coverage measurements handle that second structure properly.

**Pull request merged**